pkg/source/github/github_test.go

```diff
@@ -0,0 +1,373 @@
+				// Setup
+				// Reset the BaseURL to force an error
+				src.client.BaseURL, err = url.Parse("invalidurl")
+				Expect(err).ToNot(HaveOccurred())
+				
+				// Run test
+				_, err := src.FetchRelease(releaseTag)
+
+				// Validate results
+				Expect(err).To(HaveOccurred())
+			})
+		})
+	})
+
+	When("FetchLatestRelease() is called and", func() {
+		When("at lease one release exists", func() {
+			It("returns the release without error", func() {
+				// Setup
+				respTag := "test"
+				resp := gogithub.RepositoryRelease {
+					TagName: &respTag,
+				}
+				src.AddFetchLatestReleaseResponse(resp)
+
+				// Run test
+				result, err := src.FetchLatestRelease()
+
+				// Validate results
+				Expect(err).ToNot(HaveOccurred())
+				Expect(*result).To(Equal(resp))
+			})
+		})
+		When("no releases exist for the source", func() {
+			It("returns an error to handle", func() {
+				// Run test
+				_, err := src.FetchLatestRelease()
+
+				// Validate results
+				Expect(err).To(HaveOccurred())
+			})
+		})
+		When("the go-github client returns an error", func() {
+			It("is returned so it can be handled", func() {
+				// Setup
+				// Reset the BaseURL to force an error
+				src.client.BaseURL, err = url.Parse("invalidurl")
+				Expect(err).ToNot(HaveOccurred())
+				
+				// Run test
+				_, err := src.FetchLatestRelease()
+
+				// Validate results
+				Expect(err).To(HaveOccurred())
+			})
+		})
+	})
+
+	When("DownloadReleaseAssets() is called and", func() {
+		var (
+			asset1, asset2, asset3 *gogithub.ReleaseAsset
+
+			asset1Tag = int64(123)
+			asset2Tag = int64(456)
+			asset3Tag = int64(789)
+
+			asset1Name = "asset1"
+			asset2Name = "asset2"
+			asset3Name = "asset3"
+
+			assets []*gogithub.ReleaseAsset
+
+			tmp string
+		)
+		BeforeEach(func() {
+			// Create each test asset, with the contents of each asset being set as their name
+			asset1 = &gogithub.ReleaseAsset{
+				Name: &asset1Name,
+				ID: &asset1Tag,
+			}
+			err = src.AddDownloadReleaseAssetResponse(asset1, asset1Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			asset2 = &gogithub.ReleaseAsset{
+				Name: &asset2Name,
+				ID: &asset2Tag,
+			}
+			err = src.AddDownloadReleaseAssetResponse(asset2, asset2Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			asset3 = &gogithub.ReleaseAsset{
+				Name: &asset3Name,
+				ID: &asset3Tag,
+			}
+			err = src.AddDownloadReleaseAssetResponse(asset3, asset3Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			// Create asset slice and tmp dir for the following tests
+			assets = []*gogithub.ReleaseAsset{asset1, asset2, asset3}
+
+			tmp, err = os.MkdirTemp(os.TempDir(), "")
+			Expect(err).ToNot(HaveOccurred())
+		})
+		// Clean the tmp dir and un-populate the assets slice after each test to avoid cross-contamination and memory leaks
+		// (Note: Changes to src are automatically handled at the Context level, so no direct cleanup required)
+		AfterEach(func() {
+			err = os.RemoveAll(tmp)
```

Comment:
Does this run even if a test panics? While I appreciate the need to test with a temp directory in such a file-centric application, I'm just a little concerned we have the possibility to leak temp directories over time. Could we push the IO dependency as far out on the fringe as we can to reduce the amount of test cases that need to do this? I'm not sure if this is possible, but if we can get the code to "this URL goes into a file at this location" and reuse that everywhere, we might be in a better place.